### ISSUE TITLE
void foo(T& out) - How to fix output parameters

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#8737]
void foo(T&amp; out) - How to fix output parameters<br>
<br>
https://ift.tt/LHI7Z8b<br>
<br>
<br>
<br>
<div>
<div>
<div>
<div>
<article>
<header>
<h1>void foo(T&amp; out) - How to fix output parameters</h1>
</header>
<p>There are some cases where you need to return a value from a function but cannot use the return value. It happens, for example, in functions where you want to return multiple values at once. While you can pass multiple inputs to a function - the parameters, you cannot pass multiple return values in the same way.</p>
<p>C++ programmers tend to use a good old (lvalue) reference for that. You take a non-<code>const</code> reference as parameter and assign the output to that reference. The caller will pass a variable and upon function completion find the value of the variable changed.</p>
<p>Yet this approach has some problems: For starters, it is not obvious when just looking at the call that the variable is going to be changed. This is the reason that C++ style guides such as the one used by <a href="https://google.github.io/styleguide/cppguide.html#Reference_Arguments">Google</a> recommend using a pointer for that. The caller then has to explicitly pass in the address of the variable, making it explicit.</p>
<p>But with a pointer you can now pass in <code>nullptr</code>, you have to check for that in the function: A pointer where you really mean &ldquo;reference&rdquo; does not follow <a href="https://www.foonathan.net/blog/2016/09/27/error-handling-types.html">the guidelines</a> I&rsquo;ve been <a href="https://www.foonathan.net/blog/2016/10/11/type-safe.html">advocating for</a>.</p>
<p>So is there not a universal solution?</p>
<p>There is, but first we need to understand the full scope of the problem.</p>
<h2>Motivation</h2>
<h3>Update: Disclaimer</h3>
<p>This post does not intend to advocate for or against the use of output parameters in general. The motivation is simply here to acknowledge that output parameters are a thing people might want to use. Then I go on and show how they can be implemented in a nicer way.</p>
<p>I do not want to convince you to use output parameters neither do I want to convince you to <em>not</em> use output parameters. I just want to say: If you want to use output parameters, consider using the technique described here, as it is less error prone. But if you don&rsquo;t want to use output parameters, don&rsquo;t use them.</p>
<p>Continue reading.</p>
<hr>
<p>Let&rsquo;s assume we&rsquo;re having a function <code>func()</code> where we want to return multiple values:</p>
<div>
<pre tabindex="0"><code><span><span><span>int</span> <span>func</span><span>(</span><span>int</span> <span>in_a</span><span>,</span> <span>int</span> <span>in_b</span><span>,</span> <span>int</span><span>&amp;</span> <span>out</span><span>);</span>
</span></span></code></pre>
</div>
<blockquote>
<p>The name of this function could be <code>divide()</code> and it returns result and rest.</p>
</blockquote>
<p>Using an lvalue reference for that allows calling it like so:</p>
<div>
<pre tabindex="0"><code><span><span><span>int</span> <span>output</span><span>;</span>
</span></span><span><span><span>auto</span> <span>result</span> <span>=</span> <span>func</span><span>(</span><span>42</span><span>,</span> <span>11</span><span>,</span> <span>output</span><span>);</span>
</span></span></code></pre>
</div>
<p>This has a couple of problems, though, as already mentioned:</p>
<ul>
<li>
<p>It is not obvious that <code>output</code> is going to be modified.</p>
</li>
<li>
<p><code>output</code> needs to be created first - this requires a default constructor! In this example it is not initialized at all, this can be bad if the function returns without setting a value (because of a bug in the implementation, or documented behavior)</p>
</li>
</ul>
<p>But there is another problem which occurs in the definition of the function. Consider the following functions, that concatenates all strings from a stream. The following definition can lead to a result which might be surprising:</p>
<div>
<pre tabindex="0"><code><span><span><span>bool</span> <span>read_strings</span><span>(</span><span>std</span><span>::</span><span>istream</span><span>&amp;</span> <span>in</span><span>,</span> <span>std</span><span>::</span><span>string</span><span>&amp;</span> <span>out</span><span>)</span>
</span></span><span><span><span>{</span>
</span></span><span><span> <span>for</span> <span>(</span><span>std</span><span>::</span><span>string</span> <span>tmp</span><span>;</span> <span>in</span> <span>&gt;&gt;</span> <span>tmp</span><span>;)</span>
</span></span><span><span> <span>out</span> <span>+=</span> <span>tmp</span><span>;</span>
</span></span><span><span> <span>return</span> <span>!</span><span>out</span><span>.</span><span>empty</span><span>();</span>
</span></span><span><span><span>}</span>
</span></span></code></pre>
</div>
<blockquote>
<p>I get that this example can simply return the <code>std::string</code>.</p>
<p>But because it returns <code>bool</code>, you can use it in some form of loop or similar.</p>
</blockquote>
<p><code>read_strings()</code> concatenates <code>tmp</code> with <code>out</code> by repeatedly calling <code>operator+=()</code>. This only gives the desired results if <code>out</code> was initially empty. The following call</p>
<div>
<pre tabindex="0"><code><span><span><span>std</span><span>::</span><span>string</span> <span>str</span> <span>=</span> <span>"abc"</span><span>;</span>
</span></span><span><span><span>read_strings</span><span>(</span><span>in</span><span>,</span> <span>str</span><span>);</span>
</span></span></code></pre>
</div>
<p>will keep the value of <code>abc</code> in <code>str</code>. This might be surprising behavior.</p>
<p>So even when <em>implementing</em> functions with &ldquo;naive&rdquo; output parameters, one has to be careful and not accidentally <em>read</em> from it before setting it to a known value. This is in addition to all the problems for the caller.</p>
<h2>The solution?</h2>
<blockquote>
<p>THIS IS NOT THE REAL SOLUTION, DO READ ON!</p>
</blockquote>
<p>All of this can be avoided by simply using a <code>struct</code>:</p>
<div>
<pre tabindex="0"><code><span><span><span>struct</span> <span>func_result</span>
</span></span><span><span><span>{</span>
</span></span><span><span> <span>int</span> <span>first_value</span><span>;</span>
</span></span><span><span> <span>int</span> <span>second_value</span><span>;</span>
</span></span><span><span><span>};</span>
</span></span><span><span>
</span></span><span><span><span>func_result</span> <span>func</span><span>(</span><span>int</span> <span>a</span><span>,</span> <span>int</span> <span>b</span><span>);</span>
</span></span><span><span>
</span></span><span><span><span>&hellip;</span>
</span></span><span><span>
</span></span><span><span><span>auto</span> <span>res</span> <span>=</span> <span>func</span><span>(</span><span>42</span><span>,</span> <span>11</span><span>);</span>
</span></span><span><span><span>std</span><span>::</span><span>cout</span> <span>&lt;&lt;</span> <span>res</span><span>.</span><span>first_value</span> <span>&lt;&lt;</span> <span>' '</span> <span>&lt;&lt;</span> <span>res</span><span>.</span><span>second_value</span> <span>&lt;&lt;</span> <span>'\n'</span><span>;</span>
</span></span></code></pre>
</div>
<p>Real code would use proper names for, well, <em>everything</em> shown here, but you get the point. We&rsquo;ve declared a new <em>type</em> for the return type, one that can represent and store the two values. Then we can just return that type at once.</p>
<blockquote>
<p>One &ldquo;could&rdquo; also use <code>std::pair</code>/<code>std::tuple</code> for that, but because they do not give meaningful names to their members and instead rely on the order, this is error prone. Using those in any form of public, commonly used API is not something anybody would ever, do, am I right?</p>
</blockquote>
<p>I wouldn&rsquo;t write a blog post telling you just &ldquo;use a <code>struct</code> if you want to return multiple values&rdquo;. This is also <a href="http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-out-multi">recommended by the C++ Core Guidelines</a>. Furthermore, that&rsquo;s not a solution all the time:</p>
<div>
<pre tabindex="0"><code><span><span><span>std</span><span>::</span><span>string</span> <span>a</span><span>,</span> <span>b</span><span>;</span>
</span></span><span><span><span>std</span><span>::</span><span>cin</span> <span>&gt;&gt;</span> <span>a</span> <span>&gt;&gt;</span> <span>b</span><span>;</span>
</span></span></code></pre>
</div>
<p>What are you going to do now?!</p>
<p>There are many situations where you cannot use the return type of a function. Operator overloading is the least convincing one, I can also continue with callbacks and other form of functors you pass to something and they call you back.</p>
<blockquote>
<p>See the disclaimer above: This is just here to show that the output parameter problem does not always have to do with multiple return values. You can of course always avoid output parameters, if you really want to.</p>
</blockquote>
<p>In all those cases, the return type is fixed, so you can&rsquo;t use a <code>struct</code>.</p>
<h2>The solution</h2>
<p>In a discussion about the Google style guide and that particular rule about output parameters, I heard someone - jokingly? - suggesting that one should use <code>std::ref()</code>. In case you don&rsquo;t know, <code>std::ref()</code> returns a <code>std::reference_wrapper&lt;T&gt;</code>, which is a wrapper over a <del>reference</del> pointer where assignment changes the object it refers to. It is implicitly convertible to <code>T&amp;</code>, so you could use it in our original example like so:</p>
<div>
<pre tabindex="0"><code><span><span><span>int</span> <span>output</span><span>;</span>
</span></span><span><span><span>auto</span> <span>result</span> <span>=</span> <span>func</span><span>(</span><span>42</span><span>,</span> <span>11</span><span>,</span> <span>std</span><span>::</span><span>ref</span><span>(</span><span>output</span><span>));</span>
</span></span></code></pre>
</div>
<p>But the compiler does not enforce that, so it is not as great as it could be. The next step might be to change the <em>parameter</em> type as well. What happens if we change the output parameter to <code>std::reference_wrapper</code>?</p>
<div>
<pre tabindex="0"><code><span><span><span>int</span> <span>func</span><span>(</span><span>int</span> <span>in_a</span><span>,</span> <span>int</span> <span>in_b</span><span>,</span> <span>std</span><span>::</span><span>reference_wrapper</span><span>&lt;</span><span>int</span><span>&gt;</span> <span>out</span><span>);</span>
</span></span></code></pre>
</div>
<p>This is not a great idea, because <code>std::reference_wrapper</code> does not behave like references. Assignment rebinds the reference, so we have to use <code>out.get() = &hellip;</code> instead of <code>out = &hellip;</code>. Furthermore, <code>std::reference_wrapper</code> is <em>still</em> implicitly created from a reference, so we can just pass it in without being obvious in the caller.</p>
<p>But changing the parameter type is a step in the right direction. We just need a <em>new</em> type - <code>output_parameter&lt;T&gt;</code>. This type should have an <code>explicit</code> constructor taking <code>T&amp;</code> which stores a pointer to the object. Then it needs an assignment operator that takes some <code>T</code> and assigns that to the pointer:</p>
<div>
<pre tabindex="0"><code><span><span><span>template</span> <span>&lt;</span><span>typename</span> <span>T</span><span>&gt;</span>
</span></span><span><span><span>class</span> <span>output_parameter</span>
</span></span><span><span><span>{</span>
</span></span><span><span><span>public</span><span>:</span>
</span></span><span><span> <span>explicit</span> <span>output_parameter</span><span>(</span><span>T</span><span>&amp;</span> <span>obj</span><span>)</span>
</span></span><span><span> <span>:</span> <span>ptr_</span><span>(</span><span>&amp;</span><span>obj</span><span>)</span> <span>{}</span>
</span></span><span><span>
</span></span><span><span> <span>// disable assignment because it would have totally different semantics then the other operator=
</span></span></span><span><span> <span>// also it is really not needed
</span></span></span><span><span> <span>output_parameter</span><span>&amp;</span> <span>operator</span><span>=</span><span>(</span><span>const</span> <span>output_parameter</span><span>&amp;</span><span>)</span> <span>=</span> <span>delete</span><span>;</span>
</span></span><span><span>
</span></span><span><span> <span>output_parameter</span><span>&amp;</span> <span>operator</span><span>=</span><span>(</span><span>T</span> <span>value</span><span>)</span>
</span></span><span><span> <span>{</span>
</span></span><span><span> <span>*</span><span>ptr_</span> <span>=</span> <span>std</span><span>::</span><span>move</span><span>(</span><span>value</span><span>);</span>
</span></span><span><span> <span>return</span> <span>*</span><span>this</span><span>;</span>
</span></span><span><span> <span>}</span>
</span></span><span><span>
</span></span><span><span><span>private</span><span>:</span>
</span></span><span><span> <span>T</span><span>*</span> <span>ptr_</span><span>;</span>
</span></span><span><span><span>};</span>
</span></span></code></pre>
</div>
<p>We make it the parameter type and now a caller has to write:</p>
<div>
<pre tabindex="0"><code><span><span><span>int</span> <span>output</span><span>;</span>
</span></span><span><span><span>auto</span> <span>result</span> <span>=</span> <span>func</span><span>(</span><span>42</span><span>,</span> <span>11</span><span>,</span> <span>output_parameter</span><span>&lt;</span><span>int</span><span>&gt;</span><span>(</span><span>output</span><span>));</span>
</span></span></code></pre>
</div>
<p>Hm, that might be <em>too</em> verbose. No problem, simply use a helper function:</p>
<div>
<pre tabindex="0"><code><span><span><span>int</span> <span>output</span><span>;</span>
</span></span><span><span><span>auto</span> <span>result</span> <span>=</span> <span>func</span><span>(</span><span>42</span><span>,</span> <span>11</span><span>,</span> <span>out</span><span>(</span><span>output</span><span>));</span>
</span></span></code></pre>
</div>
<p>Now it is obvious that <code>output</code> is an output parameter and modified from the call. Furthermore, you cannot pass <code>output</code> directly, so it is enforced by the compiler</p>
<ul>
<li>first disadvantage eliminated.</li>
</ul>
<p>Let&rsquo;s look at the <code>read_strings()</code> implementation again:</p>
<div>
<pre tabindex="0"><code><span><span><span>bool</span> <span>read_strings</span><span>(</span><span>std</span><span>::</span><span>istream</span><span>&amp;</span> <span>in</span><span>,</span> <span>output_parameter</span><span>&lt;</span><span>std</span><span>::</span><span>string</span><span>&gt;</span> <span>out</span><span>)</span>
</span></span><span><span><span>{</span>
</span></span><span><span> <span>std</span><span>::</span><span>string</span> <span>result</span><span>;</span>
</span></span><span><span> <span>for</span> <span>(</span><span>std</span><span>::</span><span>string</span> <span>tmp</span><span>;</span> <span>in</span> <span>&gt;&gt;</span> <span>tmp</span><span>;)</span>
</span></span><span><span> <span>result</span> <span>+=</span> <span>tmp</span><span>;</span>
</span></span><span><span> <span>out</span> <span>=</span> <span>std</span><span>::</span><span>move</span><span>(</span><span>result</span><span>);</span>
</span></span><span><span> <span>return</span> <span>!</span><span>result</span><span>.</span><span>empty</span><span>();</span>
</span></span><span><span><span>}</span>
</span></span></code></pre>
</div>
<p>Because we can&rsquo;t use <code>operator+=</code> on <code>out</code>, we have to use a temporary string and move that in: We can&rsquo;t accidentally read from <code>out</code>. But this implementation has a bug - after the <code>std::move()</code>, <code>result</code> might be empty. So we have to get the result first:</p>
<div>
<pre tabindex="0"><code><span><span><span>bool</span> <span>read_strings</span><span>(</span><span>std</span><span>::</span><span>istream</span><span>&amp;</span> <span>in</span><span>,</span> <span>output_parameter</span><span>&lt;</span><span>std</span><span>::</span><span>string</span><span>&gt;</span> <span>out</span><span>)</span>
</span></span><span><span><span>{</span>
</span></span><span><span> <span>std</span><span>::</span><span>string</span> <span>result</span><span>;</span>
</span></span><span><span> <span>for</span> <span>(</span><span>std</span><span>::</span><span>string</span> <span>tmp</span><span>;</span> <span>in</span> <span>&gt;&gt;</span> <span>tmp</span><span>;)</span>
</span></span><span><span> <span>result</span> <span>+=</span> <span>tmp</span><span>;</span>
</span></span><span><span> <span>auto</span> <span>empty</span> <span>=</span> <span>result</span><span>.</span><span>empty</span><span>();</span>
</span></span><span><span> <span>out</span> <span>=</span> <span>std</span><span>::</span><span>move</span><span>(</span><span>result</span><span>);</span>
</span></span><span><span> <span>return</span> <span>!</span><span>empty</span><span>;</span>
</span></span><span><span><span>}</span>
</span></span></code></pre>
</div>
<p>Granted, that&rsquo;s verbose.</p>
<p>We want to prevent reading from <code>out</code> before we know the value. If we just add a <code>get_value()</code> function or similar, this is not statically checked. So we need to make the value available only <em>after</em> we have assigned it.</p>
<p>How can we do that?</p>
<p>Simple: just change the return type of the assignment operator. It is <code>T&amp; T::operator=(&hellip;)</code> by convention and to allow <code>a = b = c</code>. But our assignment operator does not really behave like a regular assignment operator, so there is no harm in changing that convention. We can thus change the return <em>type</em>: the only disadvantage we have is removing the ability to do <code>a = b = c</code>, but what would the semantics be anyways?</p>
<blockquote>
<p>This only works here because <code>output_parameter</code> is not meant to be used in generic code. It&rsquo;s only use is as a function parameter, where the type is clearly spelled out.</p>
</blockquote>
<p>So, let&rsquo;s change the signature of <code>output_parameter::operator=()</code>:</p>
<div>
<pre tabindex="0"><code><span><span><span>T</span><span>&amp;</span> <span>operator</span><span>=</span><span>(</span><span>T</span> <span>value</span><span>)</span>
</span></span><span><span><span>{</span>
</span></span><span><span> <span>*</span><span>ptr_</span> <span>=</span> <span>std</span><span>::</span><span>move</span><span>(</span><span>value</span><span>);</span>
</span></span><span><span> <span>return</span> <span>*</span><span>ptr_</span><span>;</span>
</span></span><span><span><span>}</span>
</span></span></code></pre>
</div>
<p>I&rsquo;ve changed the return type to <code>T&amp;</code> and let it return the value. This is exactly what we want: We can get the value, but only after we know that it is in a known state! There is no way to <em>get</em> the value without assigning it because we can only get it, after we&rsquo;ve assigned it!</p>
<p>With that our implementation of <code>read_strings()</code> can now look like this:</p>
<div>
<pre tabindex="0"><code><span><span><span>bool</span> <span>read_strings</span><span>(</span><span>std</span><span>::</span><span>istream</span><span>&amp;</span> <span>in</span><span>,</span> <span>output_parameter</span><span>&lt;</span><span>std</span><span>::</span><span>string</span><span>&gt;</span> <span>out</span><span>)</span>
</span></span><span><span><span>{</span>
</span></span><span><span> <span>std</span><span>::</span><span>string</span> <span>result</span><span>;</span>
</span></span><span><span> <span>for</span> <span>(</span><span>std</span><span>::</span><span>string</span> <span>tmp</span><span>;</span> <span>in</span> <span>&gt;&gt;</span> <span>tmp</span><span>;)</span>
</span></span><span><span> <span>result</span> <span>+=</span> <span>tmp</span><span>;</span>
</span></span><span><span> <span>return</span> <span>!</span><span>(</span><span>out</span> <span>=</span> <span>std</span><span>::</span><span>move</span><span>(</span><span>result</span><span>)).</span><span>empty</span><span>();</span>
</span></span><span><span><span>}</span>
</span></span></code></pre>
</div>
<p>We call <code>empty()</code> on the result of the assignment operator which is the value of our output type!</p>
<p>But now we have to create two strings and have the cost of a move assign. Can it be improved?</p>
<p>Sure, just change the implementation:</p>
<div>
<pre tabindex="0"><code><span><span><span>bool</span> <span>read_strings</span><span>(</span><span>std</span><span>::</span><span>istream</span><span>&amp;</span> <span>in</span><span>,</span> <span>output_parameter</span><span>&lt;</span><span>std</span><span>::</span><span>string</span><span>&gt;</span> <span>out</span><span>)</span>
</span></span><span><span><span>{</span>
</span></span><span><span> <span>auto</span><span>&amp;</span> <span>result</span> <span>=</span> <span>(</span><span>out</span> <span>=</span> <span>""</span><span>);</span>
</span></span><span><span> <span>for</span> <span>(</span><span>std</span><span>::</span><span>string</span> <span>tmp</span><span>;</span> <span>in</span> <span>&gt;&gt;</span> <span>tmp</span><span>;)</span>
</span></span><span><span> <span>result</span> <span>+=</span> <span>tmp</span><span>;</span>
</span></span><span><span> <span>return</span> <span>!</span><span>result</span><span>.</span><span>empty</span><span>();</span>
</span></span><span><span><span>}</span>
</span></span></code></pre>
</div>
<p>We assign <code>out</code> to the empty string directly and are then allowed to work with the output parameter. With just this class already, we&rsquo;ve completely eliminated the bug that would previously occur on:</p>
<div>
<pre tabindex="0"><code><span><span><span>std</span><span>::</span><span>string</span> <span>str</span> <span>=</span> <span>"abc"</span><span>;</span>
</span></span><span><span><span>read_strings</span><span>(</span><span>in</span><span>,</span> <span>out</span><span>(</span><span>str</span><span>));</span>
</span></span></code></pre>
</div>
<p>Now this bug cannot occur anymore - by <em>type design</em>!</p>
<blockquote>
<p>Sometimes you do want to append to an existing string though. In this case you can easily write <code>read_strings(in, std::move(str), out(str))</code>, where <code>read_strings()</code> is: <code>bool read_strings(std::istream&amp; in, std::string initial, output_parameter&lt;std::string&gt; out)</code>. The function appends on <code>initial</code> and move assigns that to <code>out</code>. This is more trouble than a simple reference though, but in-out parameters - in public (!) APIs - are rare anyways.</p>
</blockquote>
<p>We&rsquo;ve thus solved two of the problems, the only thing remaining is the default constructor requirement.</p>
<h2>Allowing non-default constructible types</h2>
<p>We still have to create the variable that will be used as output prior to the function call. This still requires a default constructor or at least some way to initialize the value beforehand. What we want is a way to just create the <em>storage</em> for the object, not the object itself. We need to represent an object that might not be there yet.</p>
<p>If you&rsquo;re thinking <code>std::optional</code> or - better - <a href="https://www.foonathan.net/blog/2016/10/11/type-safe.html#tsoptionalt-and-tsoptionalreft"><code>type_safe::optional</code></a>, you&rsquo;re almost there. An optional is a type that either has a value or none. Yes, this can be used to achieve our goal because it does not require a default constructor and we can easily augment <code>output_parameter</code> so that it can handle an optional.</p>
<p>But this is not quite the abstraction we want.</p>
<p>We do not want to introduce a null state to our variable throughout its entire lifetime. What we want is a variable where the initialization is just delayed, deferred until we can initialize it. But the important point is: once it <em>is</em> initialized, it will <em>stay</em> initialized, we should not be able to un-initialize it again</p>
<ul>
<li>this would just add unnecessary complications to the code.</li>
</ul>
<p>The answer is an optional with a reduced interface - <code>deferred_construction&lt;T&gt;</code>. Like optional it has a <code>has_value()</code> function to query if it is initialized and <code>value()</code> to return the value. But the fundamental difference is: once <code>has_value()</code> returns <code>true</code>, it <em>will</em> return true for the entire lifetime of the object, so you can safely rely on that.</p>
<p>It can be implemented using my <code>type_safe::optional</code> like so:</p>
<div>
<pre tabindex="0"><code><span><span><span>template</span> <span>&lt;</span><span>typename</span> <span>T</span><span>&gt;</span>
</span></span><span><span><span>class</span> <span>deferred_construction</span>
</span></span><span><span><span>{</span>
</span></span><span><span><span>public</span><span>:</span>
</span></span><span><span> <span>deferred_construction</span><span>()</span> <span>=</span> <span>default</span><span>;</span> <span>// creates it un-initialized
</span></span></span><span><span>
</span></span><span><span> <span>deferred_construction</span><span>(</span><span>const</span> <span>deferred_construction</span><span>&amp;</span><span>)</span> <span>=</span> <span>default</span><span>;</span>
</span></span><span><span> <span>deferred_construction</span><span>(</span><span>deferred_construction</span><span>&amp;&amp;</span><span>)</span> <span>=</span> <span>default</span><span>;</span>
</span></span><span><span>
</span></span><span><span> <span>~</span><span>deferred_construction</span><span>()</span> <span>=</span> <span>default</span><span>;</span>
</span></span><span><span>
</span></span><span><span> <span>// see below
</span></span></span><span><span> <span>deferred_construction</span><span>&amp;</span> <span>operator</span><span>=</span><span>(</span><span>const</span> <span>deferred_construction</span><span>&amp;</span><span>)</span> <span>=</span> <span>delete</span><span>;</span>
</span></span><span><span>
</span></span><span><span> <span>// initializes it
</span></span></span><span><span> <span>deferred_construction</span><span>&amp;</span> <span>operator</span><span>=</span><span>(</span><span>T</span> <span>value</span><span>)</span>
</span></span><span><span> <span>{</span>
</span></span><span><span> <span>assert</span><span>(</span><span>!</span><span>has_value</span><span>());</span> <span>// see below
</span></span></span><span><span> <span>opt_</span> <span>=</span> <span>std</span><span>::</span><span>move</span><span>(</span><span>value</span><span>);</span>
</span></span><span><span> <span>return</span> <span>*</span><span>this</span><span>;</span>
</span></span><span><span> <span>}</span>
</span></span><span><span>
</span></span><span><span> <span>// + variadic emplace(Args&amp;&amp;... args) to initialize in-place
</span></span></span><span><span>
</span></span><span><span> <span>bool</span> <span>has_value</span><span>()</span> <span>const</span>
</span></span><span><span> <span>{</span>
</span></span><span><span> <span>return</span> <span>opt_</span><span>.</span><span>has_value</span><span>();</span>
</span></span><span><span> <span>}</span>
</span></span><span><span>
</span></span><span><span> <span>// + non-const overload
</span></span></span><span><span> <span>const</span> <span>T</span><span>&amp;</span> <span>value</span><span>()</span> <span>const</span>
</span></span><span><span> <span>{</span>
</span></span><span><span> <span>return</span> <span>opt_</span><span>.</span><span>value</span><span>();</span>
</span></span><span><span> <span>}</span>
</span></span><span><span>
</span></span><span><span><span>private</span><span>:</span>
</span></span><span><span> <span>type_safe</span><span>::</span><span>optional</span><span>&lt;</span><span>T</span><span>&gt;</span> <span>opt_</span><span>;</span>
</span></span><span><span><span>};</span>
</span></span></code></pre>
</div>
<p>The implementation is straightforward, there are just two unusual design decisions.</p>
<p>First, there is no assignment operator. This is required to ensure that it cannot be un-initialized. Otherwise it would allow writing:</p>
<div>
<pre tabindex="0"><code><span><span><span>deferred_construction</span><span>&lt;</span><span>T</span><span>&gt;</span> <span>obj</span><span>;</span>
</span></span><span><span><span>obj</span> <span>=</span> <span>T</span><span>(</span><span>&hellip;</span><span>);</span>
</span></span><span><span><span>obj</span> <span>=</span> <span>deferred_construction</span><span>&lt;</span><span>T</span><span>&gt;</span><span>();</span>
</span></span></code></pre>
</div>
<p>While we can simply make that assignment a no-op or assert that <code>other</code> has a value if <code>this</code> has a value, I&rsquo;ve opted for the more drastically approach of removing it.</p>
<p>Then the <code>operator=()</code> that initialises the object requires that it has not been initialized yet. While the optional itself can handle that, I&rsquo;ve decided to prevent that. The reason is simple: once the value has been initialized, the <code>deferred_construction</code> wrapper is useless, it has done what it should have done. Then you can (and should) use <code>value()</code> directly.</p>
<p>With this in place, it is now simple to augment <code>output_parameter</code>, so that it can also accept a <code>deferred_construction</code> object and handles it. The first assignment of the output parameter should use the assignment of the deferred construction object, but if it is initialized, it should use <code>value()</code> to assign.</p>
<p>Then we can write:</p>
<div>
<pre tabindex="0"><code><span><span><span>deferred_construction</span><span>&lt;</span><span>std</span><span>::</span><span>string</span><span>&gt;</span> <span>output</span><span>;</span>
</span></span><span><span><span>read_strings</span><span>(</span><span>in</span><span>,</span> <span>out</span><span>(</span><span>output</span><span>));</span>
</span></span></code></pre>
</div>
<p>And this code behaves exactly like the very first implementation, it is just safer, more obvious and does not require a default constructor.</p>
<h2>Conclusion</h2>
<p><code>output_parameter&lt;T&gt;</code> allows &ldquo;better&rdquo; output parameters where you can&rsquo;t accidentally read the value and the call is obvious. Combined with <code>deferred_construction&lt;T&gt;</code> it allows output parameters for types that are not default constructible.</p>
<p>As you&rsquo;ve probably guessed a more sophisticated implementation of everything can be found in my <a href="https://github.com/foonathan/type_safe">type_safe</a> library.</p>
<blockquote>
<p>If you've liked this blog post, consider <a href="https://paypal.me/foonathan">donating</a> or otherwise <a href="https://jonathanmueller.dev/support-me/" target="_blank">supporting me</a>.</p>
</blockquote>
<blockquote>
<p>This blog post was written for my old blog design and ported over. If there are any issues, please <a href="https://jonathanmueller.dev">let me know</a>.</p>
</blockquote>
<footer>
<div>
<section>
<figure><img src="https://gravatar.com/avatar/9cbeac7efd2dd7b3f2235df7e45d11fc?s=80"></figure>
<div>
<p>C++ enthusiast. I write libraries.</p>
</div>
</section>
</div>
</footer>
</article>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via foonathan::&#8203;blog()<br>
<br>
September 1, 2022 at 04:45PM
